crates/evm/evm/src/inspectors/revert_diagnostic.rs

```diff
@@ -0,0 +1,132 @@
+    /// Tracks whether a failed call has been spotted or not.
+    pub reverted: bool,
+}
+
+impl RevertDiagnostic {
+    fn is_delegatecall(&self, scheme: CallScheme) -> bool {
+        matches!(
+            scheme,
+            CallScheme::DelegateCall | CallScheme::ExtDelegateCall | CallScheme::CallCode
+        )
+    }
+
+    fn is_precompile(&self, spec_id: SpecId, target: Address) -> bool {
+        let precompiles = Precompiles::new(PrecompileSpecId::from_spec_id(spec_id));
+        precompiles.contains(&target)
+    }
+
+    pub fn no_code_target_address(&self, inputs: &mut CallInputs) -> Address {
+        if self.is_delegatecall(inputs.scheme) {
+            inputs.bytecode_address
+        } else {
+            inputs.target_address
+        }
+    }
+
+    pub fn reason(&self) -> Option<DetailedRevertReason> {
+        if self.reverted {
+            if let Some((addr, scheme, _)) = self.non_contract_call {
+                let reason = if self.is_delegatecall(scheme) {
+                    DetailedRevertReason::DelegateCallToNonContract(addr)
+                } else {
+                    DetailedRevertReason::CallToNonContract(addr)
+                };
+
+                return Some(reason);
+            }
+        }
+
+        if let Some((addr, _)) = self.non_contract_size_check {
+            // call never took place, so schema is unknown --> output most generic msg
+            return Some(DetailedRevertReason::CallToNonContract(addr));
+        }
+
+        None
+    }
+}
+
+impl<DB: Database + DatabaseExt> Inspector<DB> for RevertDiagnostic {
+    /// Tracks the first call, with non-zero calldata, that targeted a non-contract address.
+    /// Excludes precompiles and test addresses.
+    fn call(&mut self, ctx: &mut EvmContext<DB>, inputs: &mut CallInputs) -> Option<CallOutcome> {
+        let target = self.no_code_target_address(inputs);
+
+        if IGNORE.contains(&target) || self.is_precompile(ctx.spec_id(), target) {
+            return None;
+        }
+
+        if let Ok(state) = ctx.code(target) {
+            if state.is_empty() && !inputs.input.is_empty() && !self.reverted {
+                self.non_contract_call = Some((target, inputs.scheme, ctx.journaled_state.depth()));
+            }
+        }
+        None
+    }
+
+    /// Tracks `EXTCODESIZE` targeted to addresses without code. Clears the cache when the call
+    /// stack depth changes.
+    fn step(&mut self, interpreter: &mut Interpreter, ctx: &mut EvmContext<DB>) {
+        if let Some((_, depth)) = self.non_contract_size_check {
+            if depth != ctx.journaled_state.depth() {
+                self.non_contract_size_check = None;
+            }
+
+            return;
+        }
+
+        if EXTCODESIZE == interpreter.current_opcode() {
+            if let Ok(word) = interpreter.stack().peek(0) {
+                let addr = Address::from_word(word.into());
+                if let Ok(state) = ctx.code(addr) {
```

Comment:
this warms the account (`load_account`) hence the gas changes in tests, we should probably just record these addresses and depth and check for the code in call_end, if call reverted